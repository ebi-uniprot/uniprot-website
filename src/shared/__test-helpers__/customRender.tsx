/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
