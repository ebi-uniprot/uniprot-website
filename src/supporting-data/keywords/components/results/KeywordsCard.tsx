import { Card } from 'franklin-sites';
import { useMemo } from 'react';
import { Link } from 'react-router';

import { getEntryPath } from '../../../../app/config/urls';
import CardCheckboxCell from '../../../../shared/components/CardCheckboxCell';
import { mapToLinks } from '../../../../shared/components/MapTo';
import RenderColumnsInCard from '../../../../shared/components/results/RenderColumnsInCard';
import renderColumnsInCardStyles from '../../../../shared/components/results/styles/render-columns-in-card.module.scss';
import { Namespace } from '../../../../shared/types/namespaces';
import { getIdKeyForNamespace } from '../../../../shared/utils/getIdKey';
import { KeywordsAPIModel } from '../../adapters/keywordsConverter';
import KeywordsColumnConfiguration, {
  KeywordsColumn,
} from '../../config/KeywordsColumnConfiguration';

const category = KeywordsColumnConfiguration.get(KeywordsColumn.category);

const getIdKey = getIdKeyForNamespace(Namespace.keywords);

const KeywordsCard = ({ data }: { data: KeywordsAPIModel }) => {
  const id = getIdKey(data);

  const links = useMemo(
    () => mapToLinks(Namespace.keywords, id, data.statistics),
    [data.statistics, id]
  );

  return (
    <Card
      header={
        <>
          <CardCheckboxCell id={id} />
          <h2 className="small">
            <Link to={getEntryPath(Namespace.keywords, id)}>
              {data.keyword.name}
            </Link>
          </h2>
        </>
      }
      headerSeparator={false}
      links={links}
    >
      <div className={renderColumnsInCardStyles['result-card__info-container']}>
        {data.definition}
      </div>
      <RenderColumnsInCard renderers={category} data={data} />
    </Card>
  );
};

export default KeywordsCard;
