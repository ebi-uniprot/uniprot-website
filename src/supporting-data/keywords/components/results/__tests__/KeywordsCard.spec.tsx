import { fireEvent, screen, within } from '@testing-library/react';

import customRender from '../../../../../shared/__test-helpers__/customRender';

import KeywordsCard from '../KeywordsCard';

import keywordsData from '../../../__mocks__/keywordsModelData';

describe('KeywordsCard tests', () => {
  it('should render the card component', () => {
    const { asFragment } = customRender(
      <KeywordsCard data={keywordsData[0]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should allow card navigation', () => {
    const { history } = customRender(<KeywordsCard data={keywordsData[0]} />);
    fireEvent.click(within(screen.getByRole('heading')).getByRole('link'));
    expect(history.location.pathname).toMatch('/keywords/KW-0021');
  });
});
