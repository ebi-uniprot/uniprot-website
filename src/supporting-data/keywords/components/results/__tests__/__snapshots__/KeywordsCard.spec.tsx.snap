// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeywordsCard tests should render the card component 1`] = `
<DocumentFragment>
  <section
    class="card card--has-link"
  >
    <a
      aria-hidden="true"
      class="card__link"
      data-testid="background-link"
      href="/keywords/KW-0021"
      tabindex="-1"
    />
    <div
      class="card__container"
    >
      <div
        class="card__header"
      >
        <h2
          class="tiny"
        >
          Allosteric enzyme
        </h2>
      </div>
      <div
        class="card__content"
      >
        <div
          class="result-card__info-container"
        >
          Enzyme whose activity is modified by the noncovalent binding of an allosteric effector at a site other than the active site. This binding mediates conformational changes, altering its catalytic or binding properties.
        </div>
        <div
          class="result-card__info-container"
        >
          <span
            class="result-card__info-bit"
          >
            <strong>
              Category: 
            </strong>
            <a
              href="/keywords/KW-9992"
            >
              Molecular function
            </a>
          </span>
        </div>
      </div>
    </div>
    <div
      class="card__actions"
    >
      <a
        class="card-action"
        href="/uniprotkb?query=(keyword:KW-0021)"
      >
        142,292 UniProtKB entries
      </a>
      <a
        class="card-action"
        href="/uniprotkb?facets=reviewed:true&query=(keyword:KW-0021)"
      >
        2,482 reviewed UniProtKB entries
      </a>
      <a
        class="card-action"
        href="/uniprotkb?facets=reviewed:false&query=(keyword:KW-0021)"
      >
        139,810 unreviewed UniProtKB entries
      </a>
    </div>
  </section>
</DocumentFragment>
`;
